core/src/main/java/org/elasticsearch/watcher/DirectoryWatcher.java

```diff
@@ -0,0 +1,257 @@
+ */
+public class DirectoryWatcher extends AbstractResourceWatcher<FileChangesListener> {
+
+    private Path file;
+    private final Map<WatchKey, Path> keys;
+
+    private WatchService watchService;
+    private WatchKey rootKey;
+
+    private static final Logger logger = Loggers.getLogger(DirectoryWatcher.class);
+
+    @SuppressWarnings("unchecked")
+    static <T> WatchEvent<T> cast(WatchEvent<?> event) {
+        return (WatchEvent<T>) event;
+    }
+
+    public DirectoryWatcher(Path file) throws IOException {
+        this.file = file;
+        this.watchService = FileSystems.getDefault().newWatchService();
+        this.keys = new HashMap<WatchKey, Path>();
+
+    }
+
+    @Override
+    protected void doInit() throws IOException {
+        rootKey = file.register(watchService, ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY);
+        if (!Files.isRegularFile(file)) {
+            registerAll(file, true);
+        }
+    }
+
+    @Override
+    protected void doCheckAndNotify() throws IOException {
+        while (true) {
+            // wait for key to be signalled
+            WatchKey key;
+            try {
+                key = watchService.take();
+            } catch (InterruptedException x) {
+                return;
+            }
+
+            Path dir = keys.get(key);
+            boolean isRoot = key.equals(rootKey);
+
+            if (dir == null && !isRoot) {
+                System.err.println("WatchKey not recognized!!");
+                continue;
+            }
+
+            for (WatchEvent<?> event : key.pollEvents()) {
+                WatchEvent.Kind kind = event.kind();
+
+                // TBD - provide example of how OVERFLOW event is handled
+                if (kind == OVERFLOW) {
+                    continue;
+                }
+
+                // Context for directory entry event is the file name of entry
+                WatchEvent<Path> ev = cast(event);
+                Path name = ev.context();
+
+                if (isRoot) {
+                    if (file.getFileName().equals(name) && kind == ENTRY_DELETE) {
+                        // We only really have to worry about if this directory is deleted
+                        onDirectoryDeleted(file);
+                    }
+                } else {
+                    Path child = dir.resolve(name);
+
+                    // if directory is created, and watching recursively, then
+                    // register it and its sub-directories
+                    if (kind == ENTRY_CREATE) {
+                        try {
+                            if (Files.isDirectory(child)) {
+                                registerAll(child, false);
+                                onDirectoryCreated(child);
+                            } else {
+                                onFileCreated(child);
+                            }
+                        } catch (IOException e) {
+                            e.printStackTrace();
```

Comment:
Similar to here, gradle check will also complain about e.printStackTrace(),  make sure to fix up the other occurrences too